sqlmesh/lsp/main.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
+"""A Language Server Protocol (LSP) server for SQL with SQLMesh integration."""
+
+import logging
+import typing as t
+from contextlib import suppress
+from pathlib import Path
+
+from sqlmesh.core.context import Context
+from lsprotocol import types
+from pygls.server import LanguageServer
+from pygls.workspace import TextDocument
+from sqlmesh._version import __version__
+from sqlmesh.core.dialect import Audit
+from sqlmesh.core.model import Model
+
+logger = logging.getLogger(__name__)
+
+GLOBAL_CONTEXT: t.Optional[
+    t.Tuple[
+        Context,
+        t.Dict[str, t.Tuple[Context, t.Union[Model, Audit]]],
+    ]
+] = None
+
+
+server = LanguageServer("sqlmesh_lsp", __version__)
+
+
+def ensure_context_for_document(document: TextDocument) -> TextDocument:
+    """Ensure that a context exists for the given document if applicable by searching for a config.py or config.yml file in the parent directories."""
+    # If the context is already loaded, return the document, if it is part of the same context
+    global GLOBAL_CONTEXT
```

Comment:
why are we using global variables here with lazy initialization?

Reply:
I agree with this. Why can't we have a proper class with a state instead?